File: go/pkg/dtrules/decisiontable/anode.go
// Copyright 2004-2011 DTRules.com, Inc.
// Copyright 2024 Paul Snow
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package decisiontable

import (
	"fmt"
	"sort"

	"github.com/PaulSnow/DTRules/go/pkg/dtrules"
)

// ANode executes a list of actions.
// It represents a leaf node in the decision tree.
type ANode struct {
	decisionTable   *RDecisionTable           // The Decision Table to which this ANode belongs
	actions         []dtrules.Object          // The compiled action code to execute (Go interpreter)
	actionBytecodes []*dtrules.BytecodeChunk  // Compiled bytecode for actions (for ASM execution)
	actionNumbers   []int                     // The action numbers (0-based, for tracing)
	columns         []int                     // Column numbers that lead to this node (1-based)
	star            bool                      // Whether this column has a star
}

// SetActionBytecodes sets the bytecode for all actions.
func (a *ANode) SetActionBytecodes(bcs []*dtrules.BytecodeChunk) {
	a.actionBytecodes = bcs
}

// GetActionBytecodes returns the bytecode for all actions.
func (a *ANode) GetActionBytecodes() []*dtrules.BytecodeChunk {
	return a.actionBytecodes
}

// NewANode creates a new empty ANode for the given decision table
func NewANode(dt *RDecisionTable) *ANode {
	return &ANode{
		decisionTable: dt,
		actions:       make([]dtrules.Object, 0),
		actionNumbers: make([]int, 0),
		columns:       make([]int, 0),
	}
}

// NewANodeForColumn creates an ANode containing all actions marked with "x"
// for the given column in the decision table.
func NewANodeForColumn(dt *RDecisionTable, col int) *ANode {
	actions := make([]dtrules.Object, 0)
	numbers := make([]int, 0)

	for i := 0; i < len(dt.actionTable); i++ {
		if col < len(dt.actionTable[i]) && equalsIgnoreCase(dt.actionTable[i][col], "x") {
			if dt.ractions != nil && i < len(dt.ractions) {
				actions = append(actions, dt.ractions[i])
				numbers = append(numbers, i)
			}
		}
	}

	anode := &ANode{
		decisionTable: dt,
		actions:       actions,
		actionNumbers: numbers,
		columns:       []int{col + 1}, // 1-based column number
	}
	return anode
}

// GetRow returns -1 for action nodes (they don't correspond to condition rows)
func (a *ANode) GetRow() int {
	return -1
}

// CloneDTNode creates a deep copy of this ANode
func (a *ANode) CloneDTNode() DTNode {
	newNode := &ANode{
		decisionTable: a.decisionTable,
		actions:       make([]dtrules.Object, len(a.actions)),
		actionNumbers: make([]int, len(a.actionNumbers)),
		columns:       make([]int, len(a.columns)),
		star:          a.star,
	}
	copy(newNode.actions, a.actions)
	copy(newNode.actionNumbers, a.actionNumbers)
	copy(newNode.columns, a.columns)
	return newNode
}

// CountColumns returns 1 since an ANode represents a single endpoint
func (a *ANode) CountColumns() int {
	return 1
}

// Execute runs all actions in this node
func (a *ANode) Execute(state dtrules.State) error {
	for i, action := range a.actions {
		num := a.actionNumbers[i]

		// Save current section info
		section := state.GetCurrentTableSection()
		numHld := state.GetNumberInSection()

		// Set section to this action
		state.SetCurrentTableSection("Action", num)

		// Execute the action
		if err := state.Evaluate(action); err != nil {
			// Restore section and return error with context
			state.SetCurrentTableSection(section, numHld)
			return fmt.Errorf("action %d in table %s: %w", num+1, a.decisionTable.GetName(), err)
		}

		// Restore section
		state.SetCurrentTableSection(section, numHld)
	}
	return nil
}

// Validate always returns nil for ANodes (they are always valid leaf nodes)
func (a *ANode) Validate() *Coordinate {
	return nil
}

// EqualsNode returns true if the other node executes exactly the same actions
func (a *ANode) EqualsNode(state dtrules.State, node DTNode) bool {
	other := node.GetCommonANode(state)
	if other == nil {
		return false
	}
	if len(other.actionNumbers) != len(a.actionNumbers) {
		return false
	}
	for i := 0; i < len(a.actionNumbers); i++ {
		if other.actionNumbers[i] != a.actionNumbers[i] {
			return false
		}
	}
	return true
}

// GetCommonANode returns this node (an ANode has only one path)
func (a *ANode) GetCommonANode(state dtrules.State) *ANode {
	return a
}

// AddNode combines another node's actions into this one
// Returns error if the nodes are incompatible types
func (a *ANode) AddNode(node DTNode) error {
	other, ok := node.(*ANode)
	if !ok {
		return dtrules.TypeCheckError("ANode.AddNode", "cannot combine ANode with different node type")
	}

	// Add columns (sorted, no duplicates)
	for _, col := range other.columns {
		found := false
		for _, c := range a.columns {
			if c == col {
				found = true
				break
			}
		}
		if !found {
			a.columns = append(a.columns, col)
		}
	}
	sort.Ints(a.columns)

	// Add actions (sorted by action number, no duplicates)
	for i, idx := range other.actionNumbers {
		found := false
		for _, num := range a.actionNumbers {
			if num == idx {
				found = true
				break
			}
		}
		if !found {
			// Insert in sorted order
			pos := 0
			for pos < len(a.actionNumbers) && a.actionNumbers[pos] < idx {
				pos++
			}
			// Insert at position
			a.actionNumbers = append(a.actionNumbers, 0)
			copy(a.actionNumbers[pos+1:], a.actionNumbers[pos:])
			a.actionNumbers[pos] = idx

			a.actions = append(a.actions, nil)
			copy(a.actions[pos+1:], a.actions[pos:])
			a.actions[pos] = other.actions[i]
		}
	}
	return nil
}

// GetStar returns whether this column has a star
func (a *ANode) GetStar() bool {
	return a.star
}

// SetStar sets the star flag
func (a *ANode) SetStar(star bool) {
	a.star = star
}

// GetColumns returns the list of column numbers that lead to this node
func (a *ANode) GetColumns() []int {
	return a.columns
}

// GetDecisionTable returns the decision table this node belongs to
func (a *ANode) GetDecisionTable() *RDecisionTable {
	return a.decisionTable
}

// String returns a string representation of this ANode
func (a *ANode) String() string {
	return fmt.Sprintf("Action Node for columns %v", a.columns)
}

// Helper function for case-insensitive comparison
func equalsIgnoreCase(a, b string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := 0; i < len(a); i++ {
		ca := a[i]
		cb := b[i]
		if ca >= 'A' && ca <= 'Z' {
			ca += 'a' - 'A'
		}
		if cb >= 'A' && cb <= 'Z' {
			cb += 'a' - 'A'
		}
		if ca != cb {
			return false
		}
	}
	return true
}

File: legacy/java/dtrules-engine/src/main/java/com/dtrules/decisiontables/ANode.java
/** 
 * Copyright 2004-2011 DTRules.com, Inc.
 * 
 * See http://DTRules.com for updates and documentation for the DTRules Rules Engine  
 *   
 * Licensed under the Apache License, Version 2.0 (the "License");  
 * you may not use this file except in compliance with the License.  
 * You may obtain a copy of the License at  
 *   
 *      http://www.apache.org/licenses/LICENSE-2.0  
 *   
 * Unless required by applicable law or agreed to in writing, software  
 * distributed under the License is distributed on an "AS IS" BASIS,  
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
 * See the License for the specific language governing permissions and  
 * limitations under the License.  
 **/


package com.dtrules.decisiontables;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;

import com.dtrules.decisiontables.RDecisionTable.Type;
import com.dtrules.infrastructure.RulesException;
import com.dtrules.interpreter.IRObject;
import com.dtrules.interpreter.RArray;
import com.dtrules.interpreter.RName;
import com.dtrules.interpreter.operators.RControl.PolicyStatements;
import com.dtrules.interpreter.operators.ROperator;
import com.dtrules.session.DTState;

/**
 * ANodes execute a list of actions.
 * @author ps24876
 *
 */
public class ANode implements DTNode {
    RDecisionTable       rDecisionTable;                       // The Decision Table to which this ANode belongs
    ArrayList<IRObject>  action   = new ArrayList<IRObject>(); // The action postfix
    ArrayList<Integer>   anumbers = new ArrayList<Integer>();  // The action numbers (for tracing purposes)
    ArrayList<Integer>   columns  = new ArrayList<Integer>();  // Keep track of the columns that send us here, for tracing
    String               section;                              // What section we are being called from.
    boolean              star = false;
    
    public DTNode cloneDTNode(){
        ANode newANode = new ANode(rDecisionTable);
        newANode.action.addAll(this.action);
        newANode.anumbers.addAll(this.anumbers);
        newANode.columns.addAll(this.columns);
        newANode.section = this.section;
        return newANode;
    }
    
    public int getRow(){ return -1; }
    
    /**
     * Create and return an new ANode instance that holds all of the 
     * actions whose supposed to execute if this column executes.
     * @param col (the zero based column number)
     */
    public static ANode newANode(RDecisionTable dt, int col){
       //int cnt = 0;
       ArrayList<IRObject> list    = new ArrayList<IRObject> ();
       ArrayList<Integer>  numbers = new ArrayList<Integer>();
       for(int i=0; i< dt.actiontable.length; i++){
           if(dt.actiontable[i][col].equalsIgnoreCase("x")){
        	   if(dt.ractions!=null && dt.ractions.length>=i){
                  list.add(dt.ractions[i]);
                  numbers.add(Integer.valueOf(i));
        	   }   
           }
       }
       return new ANode(dt,col+1, list, numbers);
    }   
    /**
     * Combines into this node the given node.  Used to collapse two
     * nodes when both nodes should be executed.
     */
    public void addNode(DTNode _node){
        
        if(!(_node instanceof ANode)){
            throw new RuntimeException("Shouldn't every call if Node types don't match!");
        }
        
        ANode node = (ANode)_node;

        for(Integer column : node.columns){
            if(!columns.contains(column)) columns.add(column);
            for(int i=columns.size()-1; i>0 ; i--){
                for(int j=0; j<i; j++){
                    if(columns.get(j)>columns.get(j+1)){
                        Integer hld = columns.get(j);
                        columns.set(j,columns.get(j+1));
                        columns.set(j+1,hld);
                    }
                }
            }
        }
       
        for(int i=0;i<node.anumbers.size(); i++){
            Integer index = node.anumbers.get(i);
            if(!anumbers.contains(index)){
                int v   = index.intValue();
                int pos = 0;
                while(pos< anumbers.size() && anumbers.get(pos).intValue()<v) pos++;
                anumbers.add(pos,index);
                action.add(pos,rDecisionTable.ractions[v]);
            }
        }
    }
    
    public int countColumns(){
        return 1;
    }
    
    private ANode(RDecisionTable dt, int column, ArrayList<IRObject> objects, ArrayList<Integer> numbers){
        rDecisionTable      = dt;
    	columns.add(new Integer(column));
    	action              = objects;
        anumbers            = numbers;
    }
    
    public ANode(RDecisionTable dt){
        rDecisionTable      = dt;
    }
    
    /** Give the list of columns that got us to this ANode.  Unbalanced tables
     *  can give us multiple columns in a single ANode
     * @param columns
     * @return
     */
    public String prtColumns(ArrayList<Integer> columns){
        String s="";
        for(Integer column : columns){
            s += column.toString()+" ";
        }
        return s;
    }
    
	public void execute(DTState state) throws RulesException {
        Iterator<Integer> inum = anumbers.iterator();
        if(state.testState(DTState.TRACE)){
            state.traceTagBegin("column", "n",prtColumns(columns));
        }
        ANode oldANode = state.getAnode();
        int num = 0;
        try {
            state.setAnode(this);
            if(state.testState(DTState.TRACE)){
            	if(state.testState(DTState.VERBOSE)){
	                for(IRObject v : action){
	                    num = inum.next().intValue();
	                    state.traceTagBegin("action","n",(num+1)+"");
	                        state.traceInfo("formal",rDecisionTable.getActions()[num]);
	                        int d = state.ddepth();                
	                        String section = state.getCurrentTableSection();
	                        int    numhld  = state.getNumberInSection();
	                        state.setCurrentTableSection("Action",num);
	                        state.evaluate(v);
	                        state.setCurrentTableSection(section, numhld);
	                        if(d!=state.ddepth()){
	                            state.setAnode(oldANode);
	                            throw new RulesException("data stack unbalanced","ANode Execute","Action "+(num+1)+" in table "+rDecisionTable.getDtname());
	                        }
	                    state.traceTagEnd();
	                }
                }else{
                	for(IRObject v : action){
                        num = inum.next().intValue();
                        String section = state.getCurrentTableSection();
                        int    numhld  = state.getNumberInSection();
                        state.setCurrentTableSection("Action",num);
                        state.traceTagBegin("action","n",(num+1)+"");
                        	state.evaluate(v);
                        state.traceTagEnd();
                        state.setCurrentTableSection(section, numhld);
                    }	
                }
            }else{
                for(IRObject v : action){
                    num = inum.next().intValue();
                    String section = state.getCurrentTableSection();
                    int    numhld  = state.getNumberInSection();
                    state.setCurrentTableSection("Action",num);
                    state.evaluate(v);
                    state.setCurrentTableSection(section, numhld);
                }  
            }
        } catch (RulesException e) {
            boolean first = e.isFirstAction(); 
            if(state.testState(DTState.TRACE)){
                if(first){
                    state.traceInfo("Error_Detected",null);
                }
                state.traceTagEnd();
            }
            e.setSection("Action",num+1);
            e.setFormal(rDecisionTable.getActions()[num]);
            state.setAnode(oldANode);
            throw e;
        } catch (Exception e){
            RulesException re = new RulesException(e.getClass().getName(), e.getStackTrace()[0].getClassName(), e.getMessage());
            re.isFirstAction(); /** Just so we note that this is the first action */
            if(state.testState(DTState.TRACE)){
                state.traceInfo("Error_Detected",null);
                state.traceTagEnd();
            }
            re.setSection("Action",num+1);
            re.setFormal(rDecisionTable.getActions()[num]);
            state.setAnode(oldANode);
            throw re;
        }

        if(state.testState(DTState.TRACE)){
            state.traceTagEnd();
        }
        state.setAnode(oldANode);
	}

	public Coordinate validate() {
		return null;
	}
    
	public String toString(){
		return "Action Node for columns "+(prtColumns(columns));
	}
	/**
     * An Action Node is equal to another DTNode if 1) it has one and
     * only one set of actions it executes regardless of the path taken, and
     * 2) if the actions this node takes are exactly the as the node provided. 
	 */
    public boolean equalsNode(DTState state, DTNode node) {
        ANode other = node.getCommonANode(state);               // Get the common path            
        if(other==null)return false;                            // No common path? Not Equal then!
        if(other.anumbers.size()!=anumbers.size()) return false;// Must be the same length.
        for(int i = 0; i< anumbers.size(); i++){
            if(!other.anumbers.get(i).equals(anumbers.get(i)))return false;  //   Make sure each action is the same action.
        }                                                                    //     If a mismatch is found, Not Equal!!!
        return true;
    }
    
    public ANode getCommonANode(DTState state) {
        return this;
    }

    public RDecisionTable getrDecisionTable() {
        return rDecisionTable;
    }

    public ArrayList<Integer> getColumns() {
        return columns;
    }

    @Override
    public boolean getStar() {
        return star;
    }

    @Override
    public void setStar(boolean star) {
        this.star = star;
        
    }
    
    
}
